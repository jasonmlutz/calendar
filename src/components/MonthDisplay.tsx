import React, { useState } from "react";
import { BsArrowRightSquare, BsArrowLeftSquare } from "react-icons/bs";

type MonthName = Readonly<
  | "January"
  | "February"
  | "March"
  | "April"
  | "May"
  | "June"
  | "July"
  | "August"
  | "September"
  | "October"
  | "November"
  | "December"
>;

interface Month {
  readonly index: number;
  readonly name: MonthName;
}

export function generateMonth(index: number): Month {
  // take an index in 0 .. 11 and return the corresponding
  // Month object
  const monthNames: MonthName[] = [
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December",
  ];
  if (index > 0) {
    return {
      index: index % 12,
      name: monthNames[index % 12],
    };
  } else {
    // we want to treat an input like -1 (or -13) as we would 11
    return generateMonth(12 + index);
  }
}

export function incrementMonth(currentMonth: Month, change: number): Month {
  // return the a new Month that is change month(s) before/after current month
  return generateMonth((currentMonth.index + change) % 12);
}

function MonthDisplay() {
  let initialMonth: Month = { index: 0, name: "January" };
  const [currentMonth, setCurrentMonth] = useState(initialMonth);
  return (
    <div id="Current-month-header" className="flex flex-row justify-around">
      <div
        id="Current-month-left-nav"
        onClick={() => {
          setCurrentMonth((currentMonth) => incrementMonth(currentMonth, -1));
        }}
      >
        <BsArrowLeftSquare />
      </div>
      <div id="Current-month-name">{currentMonth.name}</div>
      <div
        id="Current-month-left-nav"
        onClick={() => {
          setCurrentMonth((currentMonth) => incrementMonth(currentMonth, 1));
        }}
      >
        <BsArrowRightSquare />
      </div>
    </div>
  );
}

export default MonthDisplay;
