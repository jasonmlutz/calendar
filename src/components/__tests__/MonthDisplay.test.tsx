import renderer from "react-test-renderer";
import MonthDisplay, { generateMonth } from "../MonthDisplay";

it("generateMonth: properly assigns month", () => {
  expect(generateMonth(0)).toEqual({ index: 0, name: "January" });
  expect(generateMonth(8)).toEqual({ index: 8, name: "September" });
});

it("generateMonth: properly handles overages", () => {
  expect(generateMonth(12)).toEqual({ index: 0, name: "January" });
  expect(generateMonth(13)).toEqual({ index: 1, name: "February" });
});

it("generateMonth: handles negative inputs", () => {
  expect(generateMonth(-1)).toEqual({ index: 11, name: "December" });
});

it("renders correctly", () => {
  const tree = renderer.create(<MonthDisplay />).toJSON();
  expect(tree).toMatchSnapshot();
});
