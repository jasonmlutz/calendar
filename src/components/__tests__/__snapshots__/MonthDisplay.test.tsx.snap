// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="flex flex-row justify-around"
  id="Current-month-header"
>
  <div
    id="Current-month-left-nav"
  >
    previous
  </div>
  <div
    id="Current-month-name"
  >
    January
  </div>
  <div
    id="Current-month-left-nav"
  >
    next
  </div>
</div>
`;
